components/remote_settings/src/context.rs

```diff
@@ -0,0 +1,123 @@
+
+use serde_json::{Map, Value};
+
+/// Remote settings context object
+///
+/// This is used to filter the records returned. We always fetch all `records` from the
+/// remote-settings storage. Some records could have a `filter_expression`.  If this is passed in
+/// and the record has a `filter_expression`, then only returns where the expression is true will
+/// be returned.
+///
+/// See https://remote-settings.readthedocs.io/en/latest/target-filters.html for details.
+#[derive(Debug, Clone, Default, uniffi::Record)]
+pub struct RemoteSettingsContext {
+    /// The delivery channel of the application (e.g "nightly")
+    #[uniffi(default = None)]
+    pub channel: Option<String>,
+    /// User visible version string (e.g. "1.0.3")
+    #[uniffi(default = None)]
+    pub app_version: Option<String>,
+    /// String containing the XUL application app_id
+    #[uniffi(default = None)]
+    pub app_id: Option<String>,
+    /// The locale of the application during initialization (e.g. "es-ES")
+    #[uniffi(default = None)]
+    pub locale: Option<String>,
+    /// The name of the operating system (e.g. "Android", "iOS", "Darwin", "WINNT")
+    #[uniffi(default = None)]
+    pub os: Option<String>,
+    /// The user-visible version of the operating system (e.g. "1.2.3")
+    #[uniffi(default = None)]
+    pub os_version: Option<String>,
+    /// Form-factor of the device ("phone", "tablet", or "desktop")
+    #[uniffi(default = None)]
+    pub form_factor: Option<String>,
+    /// Country of the user.
+    ///
+    /// This can be populated in two ways:
+    ///   - The second component of the locale
+    ///   - By using the Mozilla geolocation service, which determines country via the user's IP.
+    ///     Each Firefox app has a different way of integrating with this.
+    #[uniffi(default = None)]
+    pub country: Option<String>,
+}
+
+impl RemoteSettingsContext {
+    /// Convert this into the `env` value for the remote settings JEXL filter
+    ///
+    /// https://remote-settings.readthedocs.io/en/latest/target-filters.html
+    pub(crate) fn into_env(self) -> Value {
+        let mut v = Map::new();
+        v.insert("channel".to_string(), self.channel.into());
+        if let Some(version) = self.app_version {
+            v.insert("version".to_string(), version.into());
+        }
+        if let Some(locale) = self.locale {
+            v.insert("locale".to_string(), locale.into());
+        }
+        if self.app_id.is_some() || self.os.is_some() || self.os_version.is_some() {
+            let mut appinfo = Map::default();
+            if let Some(os) = self.os {
+                appinfo.insert("OS".to_string(), os.into());
+            }
+            if let Some(app_id) = self.app_id {
+                appinfo.insert("ID".to_string(), app_id.into());
+            }
+            if let Some(os_version) = self.os_version {
+                appinfo.insert("osVersion".to_string(), os_version.into());
```

Comment:
Here we will have `env.appinfo.OS` and `env.appinfo.osVersion`.

In Gecko, `env.os` returns:
```
{
  "isWindows": false,
  "isMac": true,
  "isLinux": false,
  "windowsVersion": null,
  "windowsBuildNumber": null,
  "macVersion": 12,
  "darwinVersion": 24.2
}
``` 
which I don't find elegant at all. I would have preferred:
```
{
  "name": "iOS",
  "version": "1.2.3",
  "buildNumber": "44"
}
```

I will add `env.appinfo.osVersion` and these `env.os` fields in Gecko, and we can decide later whether we add `env.os` to A-S too.